### PR TITLE
Add new option to other qualifications section & make it mandatory

### PR DESCRIPTION
**Pull request by JR-G**
## Context

A levels are not part of the teacher training entry requirements so candidates do not have to enter them. A levels help providers build a more detailed picture of the person applying, allowing candidates to demonstrate qualifications in a wider range of subjects. This is significant if they do not have a degree or work experience in a subject that they’re applying for.

If a person has not entered their A levels, providers are likely to ask for them later, potentially slowing down the application process. We also know, from analysis, that candidates who add 2 or more A levels have lower levels of rejection.

## Changes proposed in this pull request

The following aspects of the application have been updated:

* The other qualifications section becomes mandatory.

  In order to complete the section without any qualifications, a candidate needs to select the ‘I do not want to add A levels or other qualifications’ option (or ‘I do not want to add any other qualifications’ if an international candidate).

  This applies to new and unsubmitted applications – but not those that have already been submitted.

  If qualifications have already been added, this option is no longer shown.

* Reviewing this section when a candidate has decided not to add any qualifications replays this answer, and allows them to change it. Additional guidance is also shown reminding a candidate why adding other qualification(s) can improve their application.

* Make ‘No’ the default for adding another qualification question.

## Guidance to review

[Prototype](https://apply-beta-prototype.herokuapp.com/application/12345)

## Link to Trello card

https://trello.com/c/YYoOD2vT/3244-dev-reorder-qualification-sections-rename-other-qualifications-section-and-make-it-mandatory

## Things to check

- [ ] This code does not rely on migrations in the same Pull Request
- [ ] If this code includes a migration adding or changing columns, it also backfills existing records for consistency
- [ ] API release notes have been updated if necessary
- [ ] This code does not rely on the addition/removal of Azure config environment variables in the same Pull Request
- [ ] New environment variables have been [added to the Azure config](https://github.com/DFE-Digital/apply-for-teacher-training/blob/master/docs/environment-variables.md#azure-hosting-devops-pipeline)

### CODE REVIEW

**Comment by paulrobertlloyd** [comment]
We should only show this inset text on the section review page, not on the application review page.

File: spec/system/candidate_interface/entering_details/candidate_entering_other_qualification_no_qualification_option_choice_spec.rb [hunk 1]
```diff
@@ -0,0 +1,152 @@
...
+    then_i_see_the_select_qualification_type_page
+
+    when_i_do_not_select_any_type_option
+    and_i_click_continue
+    then_i_see_the_qualification_type_error
+
+    when_i_select_i_do_not_want_to_add_any_a_levels
+    and_i_click_continue
+    then_i_see_a_level_advice
+    then_i_see_my_no_other_qualification_selection
```
Comment (davidgisbey) [inline spec/system/candidate_interface/entering_details/candidate_entering_other_qualification_no_qualification_option_choice_spec.rb:20]: `and_i_see_my_no_other_qualification_selection `

File: spec/system/candidate_interface/entering_details/candidate_entering_other_qualification_no_qualification_option_choice_spec.rb [hunk 2]
```diff
@@ -0,0 +1,152 @@
...
+    and_i_confirm_that_i_want_to_delete_my_additional_qualification
+    then_i_see_the_select_qualification_type_page
+    and_i_select_i_do_not_want_to_add_any_a_levels
+    and_i_click_continue
+    then_i_see_a_level_advice
+    then_i_see_my_no_other_qualification_selection
+
+    when_i_mark_this_section_as_completed
+    and_i_click_on_continue
+    and_that_the_section_is_completed
```
Comment (davidgisbey) [inline spec/system/candidate_interface/entering_details/candidate_entering_other_qualification_no_qualification_option_choice_spec.rb:42]: `then_i_see_tthat_the_section_is_completed`

File: app/forms/candidate_interface/other_qualification_type_form.rb [hunk 3]
```diff
@@ -23,9 +23,10 @@ class OtherQualificationTypeForm
     attr_accessor :qualification_type
     attr_accessor :other_uk_qualification_type
     attr_accessor :non_uk_qualification_type
+    attr_accessor :no_other_qualification
 
-    validates :qualification_type, presence: true
-    validates :qualification_type, inclusion: { in: ALL_VALID_TYPES, allow_blank: false }
+    validates :qualification_type, presence: true, unless: :no_other_qualification
+    validates :qualification_type, inclusion: { in: ALL_VALID_TYPES, allow_blank: false }, unless: :no_other_qualification
```
Comment (davidgisbey) [inline app/forms/candidate_interface/other_qualification_type_form.rb]: I think it's worth adding a few tests into the validations for this new behaviour

File: app/components/candidate_interface/other_qualifications_review_component.rb [hunk 4]
```diff
@@ -31,6 +31,14 @@ def other_qualifications_rows(qualification)
       end
     end
 
+    def no_qualification_row
+      [{
```
Comment (davidgisbey) [inline app/components/candidate_interface/other_qualifications_review_component.rb:35]: Does this need to be in an array?

Comment (JR-G) [reply app/components/candidate_interface/other_qualifications_review_component.rb:35]: Yeah I think so, otherwise this happens to the layout:
image

File: app/components/candidate_interface/other_qualifications_review_component.html.erb [hunk 5]
```diff
@@ -16,5 +16,14 @@
 <% end %>
 
 <% if @qualifications.empty? %>
-  <p class="govuk-body">No other qualifications entered.</p>
+  <% if @submitting_application && !application_form.other_qualifications_completed %>
+    <%= render(CandidateInterface::IncompleteSectionComponent.new(section: application_form.international? ? :other_qualifications_international : :other_qualifications, section_path: candidate_interface_other_qualification_type_path, error: @missing_error)) %>
```
Comment (davidgisbey) [inline app/components/candidate_interface/other_qualifications_review_component.html.erb:20]: This is quite a long line. Can we do something like 

```
  <%= render CandidateInterface::IncompleteSectionComponent.new(
    section: application_form.international? ? :other_qualifications_international : :other_qualifications,
    section_path: candidate_interface_other_qualification_type_path,
    error: @missing_error
    ) 
  %>
``` 
instead

File: app/components/candidate_interface/other_qualifications_review_component.html.erb [hunk 6]
```diff
@@ -16,5 +16,14 @@
...
+  <% elsif @submitting_application && application_form.other_qualifications_completed %>
+    <%= render(SummaryCardComponent.new(rows: no_qualification_row, editable: @editable)) %>
+  <% else %>
+    <div class="govuk-inset-text govuk-!-width-two-thirds govuk-!-margin-top-0">
+      <p>Adding A levels and other qualifications makes your application stronger. They demonstrate subject knowledge not covered in your degree or work experience.</p><p class="govuk-hint">Training providers usually ask you for them later in the process.</p>
```
Comment (davidgisbey) [inline app/components/candidate_interface/other_qualifications_review_component.html.erb:25]:
```
<p class="govuk-body">Adding A levels and other qualifications makes your application stronger. They demonstrate subject knowledge not covered in your degree or work experience.</p>

<p class="govuk-hint">Training providers usually ask you for them later in the process.</p>
```

Comment (paulrobertlloyd) [reply app/components/candidate_interface/other_qualifications_review_component.html.erb:25]: Actually, please use `govuk-body` for both paragraphs. I think this is a bug in the prototype.

File: app/controllers/candidate_interface/other_qualifications/destroy_controller.rb [hunk 7]
```diff
@@ -7,10 +7,15 @@ def confirm_destroy
     end
 
     def destroy
+      if current_application.application_qualifications.other.count == 1
```
Comment (davidgisbey) [inline app/controllers/candidate_interface/other_qualifications/destroy_controller.rb:10]:
```

current_qualification.destroy!
current_application.update!(other_qualifications_completed: false)

if current_application.application_qualifications.other.count.zero?
  redirect_to candidate_interface_other_qualification_type_path
else
  redirect_to candidate_interface_review_other_qualifications_path
end
```

File: app/presenters/candidate_interface/application_form_presenter.rb [hunk 8]
```diff
@@ -223,7 +223,7 @@ def science_gcse_added?
     end
 
     def other_qualifications_completed?
-      @application_form.other_qualifications_completed
+      @application_form.other_qualifications_completed && no_incomplete_qualifications?
```
Comment (davidgisbey) [inline app/presenters/candidate_interface/application_form_presenter.rb:226]: this needs some tests please

File: app/views/candidate_interface/other_qualifications/type/_shared_form.html.erb [hunk 9]
```diff
@@ -18,6 +16,10 @@
   <%= f.govuk_radio_button :qualification_type, CandidateInterface::OtherQualificationTypeForm::NON_UK_TYPE, label: { text: 'Non-UK qualification' } do %>
     <%= f.govuk_text_field :non_uk_qualification_type, label: { text: 'Qualification name', size: 's' }, hint: { text: t('application_form.other_qualification.qualification_type.non_uk.hint_text') } %>
   <% end %>
+  <% if current_application.application_qualifications.other.blank? %>
+    <%= f.govuk_radio_divider %>
+    <%= f.govuk_radio_button :qualification_type, 'no_other_qualifications', label: { text: current_application.international? ? 'I do not want to add any other qualifications' : 'I do not want to add any A levels and other qualifications' }, hint: -> { 'Providers look at A levels and other qualifications for evidence of subject knowledge not covered in your degree or work history' } %>
```
Comment (davidgisbey) [inline app/views/candidate_interface/other_qualifications/type/_shared_form.html.erb:21]:
```
<%= f.govuk_radio_button :qualification_type, 
  'no_other_qualifications', 
  label: { text: current_application.international? ? 'I do not want to add any other qualifications' : 'I do not want to add any A levels and other qualifications' }, 
  hint: -> { 'Providers look at A levels and other qualifications for evidence of subject knowledge not covered in your degree or work history' } 
%>
```

File: app/forms/candidate_interface/other_qualification_details_form.rb [hunk 10]
```diff
@@ -70,6 +71,7 @@ def save
         )
 
       application_qualification.assign_attributes(attributes_for_persistence)
+      @current_application.update!(no_other_qualifications: false)
```
Comment (davidgisbey) [inline app/forms/candidate_interface/other_qualification_details_form.rb:74]: We should update the test for this to ensure that it covers this set to false

File: spec/system/candidate_interface/entering_details/candidate_entering_other_qualification_no_qualification_option_choice_spec.rb [hunk 11]
```diff
@@ -0,0 +1,152 @@
...
+  alias_method :and_i_choose_other, :when_i_choose_other
+
+  def when_i_fill_in_my_other_qualifications_details
+    fill_in t('application_form.other_qualification.subject.label'), with: 'History, English and Psychology'
+    fill_in t('application_form.other_qualification.grade.label'), with: 'Distinction'
+    fill_in t('application_form.other_qualification.award_year.label'), with: '2012'
+  end
+  alias_method :and_i_fill_in_my_other_qualifications_details, :when_i_fill_in_my_other_qualifications_details
+
+  def and_i_choose_not_to_add_additional_qualifications
```
Comment (davidgisbey) [inline spec/system/candidate_interface/entering_details/candidate_entering_other_qualification_no_qualification_option_choice_spec.rb:98]: Check that no is selected as default

File: spec/system/candidate_interface/entering_details/candidate_entering_other_qualification_no_qualification_option_choice_spec.rb [hunk 12]
```diff
@@ -0,0 +1,152 @@
+require 'rails_helper'
+
+RSpec.feature 'Entering their other qualifications' do
+  include CandidateHelper
+
+  scenario 'Candidate submits their other qualifications after choosing not to provide any' do
+    given_i_am_signed_in
+    and_i_visit_the_site
```
Comment (davidgisbey) [inline spec/system/candidate_interface/entering_details/candidate_entering_other_qualification_no_qualification_option_choice_spec.rb:8]: check that it now isn't complete as default

File: app/forms/candidate_interface/other_qualification_type_form.rb [hunk 13]
```diff
@@ -47,6 +47,16 @@ def save_intermediate
       valid? && save_intermediate!
     end
 
+    def save_no_other_qualifications
```
Comment (davidgisbey) [inline app/forms/candidate_interface/other_qualification_type_form.rb:50]: unit tests

File: app/components/candidate_interface/other_qualifications_review_component.html.erb [hunk 14]
```diff
@@ -16,5 +16,19 @@
 <% end %>
 
 <% if @qualifications.empty? %>
```
Comment (davidgisbey) [inline app/components/candidate_interface/other_qualifications_review_component.html.erb:18]: This needs to be tested too

File: app/forms/candidate_interface/other_qualification_type_form.rb [hunk 15]
```diff
@@ -25,7 +25,7 @@ class OtherQualificationTypeForm
     attr_accessor :non_uk_qualification_type
 
     validates :qualification_type, presence: true
-    validates :qualification_type, inclusion: { in: ALL_VALID_TYPES, allow_blank: false }
+    validates :qualification_type, inclusion: { in: ALL_VALID_TYPES, allow_blank: false }, unless: :no_other_qualification?
```
Comment (stevehook) [inline app/forms/candidate_interface/other_qualification_type_form.rb:28]: This could just be something like:

```
validates :qualification_type, inclusion: { in: ALL_VALID_TYPES + ['no_other_qualification'], allow_blank: false }
```

**Review by stevehook — COMMENTED** [review]
Added a few questions inline.

Tried the review app but for some reason the migration hadn't run. Then I was able to get to the other qualifications page for an existing candidate but I got stuck in a loop selecting 'no other qualification' redirected me straight back to the type form (https://apply-for-te-3244-dev-r-vjkqbg.herokuapp.com/candidate/application/other-qualifications)

**Comment by paulrobertlloyd** [comment]
Looking good. Just need to fix the placement of the add/add another qualification button in the different states:

When qualifications have been added, show a button link ‘Add another qualification’:

Screenshot 2021-04-15 at 14 18 51

If decided not to add any qualifications, show a button link ‘Add a qualification’ inside the inset text:

Screenshot 2021-04-15 at 14 19 35

This button will need to have `govuk-!-margin-bottom-0` class on it.

File: app/views/candidate_interface/other_qualifications/review/show.html.erb [hunk 16]
```diff
@@ -6,13 +6,13 @@
...
     </h1>
-
-    <%= govuk_link_to t('application_form.other_qualification.another.button'), candidate_interface_other_qualification_type_path, button: true, class: 'govuk-button--secondary' %>
   </div>
 </div>
 
 <%= render(CandidateInterface::OtherQualificationsReviewComponent.new(application_form: @application_form)) %>
 
+<%= govuk_button_to t('application_form.other_qualification.another.button'), candidate_interface_other_qualification_type_path, class: 'govuk-button--secondary', method: :get %>
+
```
Comment (paulrobertlloyd) [inline app/views/candidate_interface/other_qualifications/review/show.html.erb:15]: Sorry, moving this button was in the prototype, but I’ve changed its location back to where it was before, above the list of qualifications

File: app/views/candidate_interface/other_qualifications/review/show.html.erb [hunk 17]
```diff
@@ -6,13 +6,13 @@
     <h1 class="govuk-heading-xl">
```
Comment (paulrobertlloyd) [inline app/views/candidate_interface/other_qualifications/review/show.html.erb:6]: Can’t make a suggestion for all this rows… this heading and button content shouldn’t be within a grid container, so lets remove that (i.e the heading can go the full width of the page).

File: config/locales/candidate_interface/other_qualification.yml [hunk 18]
```diff
@@ -29,7 +29,7 @@ en:
         review_label: Year awarded
         change_action: year
       another:
-        button: Add another qualification
+        button: Add a qualification
```
Comment (paulrobertlloyd) [inline config/locales/candidate_interface/other_qualification.yml:32]: Revert this change.

**Review by paulrobertlloyd — CHANGES_REQUESTED** [review]
See comments.

**Comment by paulrobertlloyd** [comment]
We’re sooooo close.

Dog trying to fetch a ball in a swimming pool

Two final things:

* When choosing to ‘Add a qualification’, don’t show the option to not add a qualification.
* There appears to be a regression with the back links; both on the first page ‘What type of qualification do you want to add?’ and the subsequent qualification details page.

**Review by stevehook — APPROVED** [review]
Tested the review app and seems to be working well so I think it's ready subject to the changes requested by @paulrobertlloyd. Good stuff!

File: app/components/candidate_interface/other_qualifications_review_component.rb [hunk 19]
```diff
@@ -35,7 +35,7 @@ def no_qualification_row
       [{
         key: 'Do you want to add any A levels and other qualifications',
         value: 'No',
-        change_path: candidate_interface_other_qualification_type_path,
+        change_path: candidate_interface_other_qualification_type_path(:change_link),
```
Comment (JR-G) [inline app/components/candidate_interface/other_qualifications_review_component.rb:38]: @stevehook / @davidgisbey this feels hacky - but it does appear to be working - are there any alternatives to implement this in a cleaner way?

Comment (stevehook) [reply app/components/candidate_interface/other_qualifications_review_component.rb:38]: I think using the `format` param is bit confusing. How about a simple URL param so you end up with something like `path/to/form?no_other_qualification=true`? Maybe `candidate_interface_other_qualification_type_path(no_other_qualification: true)`?

Comment (stevehook) [reply app/components/candidate_interface/other_qualifications_review_component.rb:38]: (The type form doesn't need to know anything about change links etc. it just needs to know whether to pre-populate the radio button so the name should reflect that rather than external details)

Comment (stevehook) [reply app/components/candidate_interface/other_qualifications_review_component.rb:38]: @davidgisbey can you think of a better way to pass this switch from the review page to type form?

Comment (paulrobertlloyd) [reply app/components/candidate_interface/other_qualifications_review_component.rb]: Query string doesn’t appear to be working (happy for this to be the approach for showing/hiding the last radio option).

Can we use hyphens in the param?

And maybe the param is `?change=true`

**Review by paulrobertlloyd — APPROVED** [review]
SHIP. IT. 🎉
